From df974c57f6312d720b9a601a578393b951453c20 Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Mon, 8 Sep 2014 13:04:20 -0400
Subject: [kernel] perf: Limit perf_event_attr::sample_period to 63 bits

Message-id: <1410181752-25631-16-git-send-email-jolsa@redhat.com>
Patchwork-id: 91720
O-Subject: [PATCH RHEL7.1 BZ1134356 015/307] perf: Limit perf_event_attr::sample_period to 63 bits
Bugzilla: 1134356
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Don Zickus <dzickus@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1134356
https://bugzilla.redhat.com/show_bug.cgi?id=1134356

upstream
========
commit 0819b2e30ccb93edf04876237b6205eef84ec8d2
Author: Peter Zijlstra <peterz@infradead.org>
Date: Thu May 15 20:23:48 2014 +0200

description
===========
Vince reported that using a large sample_period (one with bit 63 set)
results in wreckage since while the sample_period is fundamentally
unsigned (negative periods don't make sense) the way we implement
things very much rely on signed logic.

So limit sample_period to 63 bits to avoid tripping over this.
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/kernel/events/core.c b/kernel/events/core.c
index c45e3d5..df00975 100644
--- a/kernel/events/core.c
+++ b/kernel/events/core.c
@@ -7033,6 +7033,9 @@ SYSCALL_DEFINE5(perf_event_open,
  if (attr.freq) {
   if (attr.sample_freq > sysctl_perf_event_sample_rate)
    return -EINVAL;
+ } else {
+  if (attr.sample_period & (1ULL << 63))
+   return -EINVAL;
  }
 
  /*
-- 
1.7.1